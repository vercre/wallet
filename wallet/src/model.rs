//! Model for the wallet application state.

pub mod credential;
mod issuance;

use chrono::{serde::ts_milliseconds_option::deserialize as ts_milliseconds_option, DateTime, Utc};
pub use credential::CredentialState;
use issuance::IssuanceState;
use serde::{Deserialize, Serialize};

use super::Aspect;

/// State for the wallet application.
#[derive(Default, Serialize)]
pub struct Model {
    /// Which aspect of the application is currently active.
    pub active_view: Aspect,

    /// Credential state.
    pub credential: CredentialState,

    /// Issuance state.
    pub issuance: Option<IssuanceState>,

    /// Error state.
    pub error: Option<String>,

    // TODO: Remove ---
    pub count: Count,
    // ----------------
}

#[derive(Serialize, Deserialize, Clone, Default, Debug, PartialEq, Eq)]
pub struct Count {
    pub value: isize,
    #[serde(deserialize_with = "ts_milliseconds_option")]
    pub updated_at: Option<DateTime<Utc>>,
}
