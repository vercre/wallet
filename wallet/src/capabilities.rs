pub mod holder_agent;
pub mod sse;
