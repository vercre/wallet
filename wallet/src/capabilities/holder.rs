//! # Wallet Holder Capability
//!
//! The holder capability provides for interactions between `vercre-holder` and
//! capabilities provided by the shell in order to implement the callbacks
//! needed. This unpacks the standard `vercre-holder` API into a more granular
//! set of helper functions that can work with Crux's event and effects system.
use std::fmt::{self, Debug};

use crux_core::capability::{CapabilityContext, Operation};
use crux_core::Capability;
use serde::{Deserialize, Serialize};
use thiserror::Error;
use vercre_holder::issuance::{OfferRequest, OfferResponse};

/// Errors that can be returned by the Holder capability.
#[derive(Clone, Debug, Deserialize, Serialize, Error, PartialEq, Eq)]
pub enum HolderError {
    /// Invalid request.
    #[error("invalid store request {message}")]
    InvalidRequest { message: String },

    /// The response from the shell capability was invalid.
    #[error("invalid store response {message}")]
    InvalidResponse { message: String },
}

/// Supported operations for the Holder capability.
#[derive(Clone, Serialize, Deserialize, PartialEq, Eq)]
pub enum HolderOperation {
    /// Process a credential offer from an issuer.
    Offer { request: OfferRequest },
}

impl Debug for HolderOperation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            HolderOperation::Offer { request } => {
                f.debug_struct("HolderOperation::Offer").field("request", request).finish()
            }
        }
    }
}

/// The result of an operation on the Holder capability.
#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq)]
pub enum HolderResult {
    /// The operation was successful.
    Ok { response: HolderResponse },

    /// The operation failed.
    Err { error: HolderError },
}

impl HolderResult {
    fn unwrap_offer(self) -> Result<OfferResponse, HolderError> {
        match self {
            HolderResult::Ok { response } => match response {
                HolderResponse::Offer { offer_response } => Ok(offer_response),
                // _ => {
                //     panic!("unexpected response type for Offer operation: {:?}", response);
                // }
            },
            HolderResult::Err { error } => Err(error.clone()),
        }
    }
}

/// The possible responses from the Holder capability.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub enum HolderResponse {
    /// Response to an offer request.
    Offer { offer_response: OfferResponse },
}

impl Operation for HolderOperation {
    type Output = HolderResult;
}

/// Capability type.
pub struct Holder<Ev> {
    context: CapabilityContext<HolderOperation, Ev>,
}

impl<Ev> Capability<Ev> for Holder<Ev> {
    type MappedSelf<MappedEv> = Holder<MappedEv>;
    type Operation = HolderOperation;

    fn map_event<F, NewEv>(&self, f: F) -> Self::MappedSelf<NewEv>
    where
        F: Fn(NewEv) -> Ev + Send + Sync + 'static,
        Ev: 'static,
        NewEv: 'static + Send,
    {
        Holder::new(self.context.map_event(f))
    }

    #[cfg(feature = "typegen")]
    fn register_types(generator: &mut crux_core::typegen::TypeGen) -> crux_core::typegen::Result {
        generator.register_type::<HolderResponse>()?;
        generator.register_type::<HolderError>()?;
        generator.register_type::<Self::Operation>()?;
        generator.register_type::<<Self::Operation as Operation>::Output>()?;
        Ok(())
    }
}

impl<Ev> Clone for Holder<Ev> {
    fn clone(&self) -> Self {
        Self {
            context: self.context.clone(),
        }
    }
}

impl<Ev> Holder<Ev>
where
    Ev: 'static,
{
    /// Creates a new Holder capability instance.
    pub fn new(context: CapabilityContext<HolderOperation, Ev>) -> Self {
        Self { context }
    }

    /// Process a credential offer from an issuer.
    pub fn offer<F>(&self, request: OfferRequest, make_event: F)
    where
        F: FnOnce(Result<OfferResponse, HolderError>) -> Ev + Send + Sync + 'static,
    {
        let context = self.context.clone();
        self.context.spawn({
            async move {
                let response = offer(&context, request).await;
                context.update_app(make_event(response))
            }
        });
    }
}

async fn offer<Ev: 'static>(
    context: &CapabilityContext<HolderOperation, Ev>, request: OfferRequest,
) -> Result<OfferResponse, HolderError> {
    context.request_from_shell(HolderOperation::Offer {
        request,
    })
    .await
    .unwrap_offer()
}
