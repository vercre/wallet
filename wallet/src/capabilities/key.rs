//! # Key Store Capability
use std::fmt::Debug;

use crux_core::capability::{CapabilityContext, Operation};
use crux_core::Capability;
use serde::{Deserialize, Serialize};
use thiserror::Error;

/// Errors that can be returned by the key store capability.
#[derive(Clone, Debug, Deserialize, Serialize, Error, PartialEq, Eq)]
pub enum KeyStoreError {
    /// Invalid request.
    #[error("invalid key store request {message}")]
    InvalidRequest { message: String },

    /// The response from the shell capability was invalid.
    #[error("invalid key store response {message}")]
    InvalidResponse { message: String },
}

/// An entry in the key store.
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub enum KeyStoreEntry {
    /// No entry for the given ID and purpose.
    None,

    /// A serialized private key.
    Data(#[serde(with = "serde_bytes")] Vec<u8>),
}

impl From<Vec<u8>> for KeyStoreEntry {
    fn from(bytes: Vec<u8>) -> Self {
        KeyStoreEntry::Data(bytes)
    }
}

impl From<KeyStoreEntry> for Option<Vec<u8>> {
    fn from(entry: KeyStoreEntry) -> Option<Vec<u8>> {
        match entry {
            KeyStoreEntry::None => None,
            KeyStoreEntry::Data(bytes) => Some(bytes),
        }
    }
}

impl From<Option<Vec<u8>>> for KeyStoreEntry {
    fn from(val: Option<Vec<u8>>) -> Self {
        match val {
            None => KeyStoreEntry::None,
            Some(bytes) => KeyStoreEntry::Data(bytes),
        }
    }
}

/// Supported operations for the key store capability.
#[derive(Clone, Serialize, Deserialize, PartialEq, Eq)]
pub enum KeyStoreOperation {
    /// Get a serialized private key from the key store.
    Get { id: String, purpose: String },

    /// Set a serialized private key in the key store.
    Set {
        id: String,
        purpose: String,
        #[serde(with = "serde_bytes")]
        data: Vec<u8>,
    },

    /// Remove a serialized private key from the key store.
    Delete { id: String, purpose: String },

    /// Generate a random secret suitable for key derivation.
    GenerateSecret { length: usize },
}

impl Debug for KeyStoreOperation {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            KeyStoreOperation::Get { id, purpose } => {
                f.debug_struct("Get").field("id", id).field("purpose", purpose).finish()
            }
            KeyStoreOperation::Set { id, purpose, data } => {
                let body = format!("<binary data - {} bytes>", data.len());
                f.debug_struct("Set")
                    .field("id", id)
                    .field("purpose", purpose)
                    .field("data", &body)
                    .finish()
            }
            KeyStoreOperation::Delete { id, purpose } => {
                f.debug_struct("Delete").field("id", id).field("purpose", purpose).finish()
            }
            KeyStoreOperation::GenerateSecret { length } => {
                f.debug_struct("GenerateSecret").field("length", length).finish()
            }
        }
    }
}

/// The possible responses from the key store capability.
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub enum KeyStoreResponse {
    /// The result of a get operation.
    Retrieved { key: KeyStoreEntry },

    /// The result of a set operation.
    Set,

    /// result of a delete operation.
    Deleted,

    /// A random secret suitable for key derivation.
    GeneratedSecret { secret: Vec<u8> },
}

/// The result of an operation on the key store.
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub enum KeyStoreResult {
    /// The operation was successful.
    Ok { response: KeyStoreResponse },

    /// The operation failed.
    Err { error: KeyStoreError },
}

impl KeyStoreResult {
    fn unwrap_get(self) -> Result<KeyStoreEntry, KeyStoreError> {
        match self {
            KeyStoreResult::Ok {
                response: KeyStoreResponse::Retrieved { key },
            } => Ok(key),
            KeyStoreResult::Err { error } => Err(error),
            _ => Err(KeyStoreError::InvalidResponse {
                message: "unexpected response for Get operation".to_string(),
            }),
        }
    }

    fn unwrap_set(self) -> Result<(), KeyStoreError> {
        match self {
            KeyStoreResult::Ok {
                response: KeyStoreResponse::Set,
            } => Ok(()),
            KeyStoreResult::Err { error } => Err(error),
            _ => Err(KeyStoreError::InvalidResponse {
                message: "unexpected response for Set operation".to_string(),
            }),
        }
    }

    fn unwrap_delete(self) -> Result<(), KeyStoreError> {
        match self {
            KeyStoreResult::Ok {
                response: KeyStoreResponse::Deleted,
            } => Ok(()),
            KeyStoreResult::Err { error } => Err(error),
            _ => Err(KeyStoreError::InvalidResponse {
                message: "unexpected response for Delete operation".to_string(),
            }),
        }
    }

    fn unwrap_generate_secret(self) -> Result<Vec<u8>, KeyStoreError> {
        match self {
            KeyStoreResult::Ok {
                response: KeyStoreResponse::GeneratedSecret { secret },
            } => Ok(secret),
            KeyStoreResult::Err { error } => Err(error),
            _ => Err(KeyStoreError::InvalidResponse {
                message: "unexpected response for GenerateSecret operation".to_string(),
            }),
        }
    }
}

impl Operation for KeyStoreOperation {
    type Output = KeyStoreResult;
}

/// Capability type for the key store.
pub struct KeyStore<Ev> {
    context: CapabilityContext<KeyStoreOperation, Ev>,
}

impl<Ev> Capability<Ev> for KeyStore<Ev> {
    type MappedSelf<MappedEv> = KeyStore<MappedEv>;
    type Operation = KeyStoreOperation;

    fn map_event<F, NewEv>(&self, f: F) -> Self::MappedSelf<NewEv>
    where
        F: Fn(NewEv) -> Ev + Send + Sync + 'static,
        Ev: 'static,
        NewEv: 'static + Send,
    {
        KeyStore::new(self.context.map_event(f))
    }

    #[cfg(feature = "typegen")]
    fn register_types(generator: &mut crux_core::typegen::TypeGen) -> crux_core::typegen::Result {
        generator.register_type::<KeyStoreResponse>()?;
        generator.register_type::<KeyStoreError>()?;
        generator.register_type::<KeyStoreEntry>()?;
        generator.register_type::<Self::Operation>()?;
        generator.register_type::<<Self::Operation as Operation>::Output>()?;
        Ok(())
    }
}

impl<Ev> Clone for KeyStore<Ev> {
    fn clone(&self) -> Self {
        Self {
            context: self.context.clone(),
        }
    }
}

impl<Ev> KeyStore<Ev>
where
    Ev: 'static,
{
    /// Create a new key store capability.
    pub fn new(context: CapabilityContext<KeyStoreOperation, Ev>) -> Self {
        Self { context }
    }

    /// Get a serialized private key from the key store and send an update event
    /// to the application.
    pub fn get<F>(
        &self, id: impl Into<String> + Send + 'static, purpose: impl Into<String> + Send + 'static,
        make_event: F,
    ) where
        F: FnOnce(Result<KeyStoreEntry, KeyStoreError>) -> Ev + Send + Sync + 'static,
    {
        self.context.spawn({
            let context = self.context.clone();
            async move {
                let response = get(&context, id, purpose).await;
                context.update_app(make_event(response))
            }
        });
    }

    /// Get a serialized private key from the key store.
    pub async fn get_async(
        &self, id: impl Into<String>, purpose: impl Into<String>,
    ) -> Result<KeyStoreEntry, KeyStoreError> {
        get(&self.context, id, purpose).await
    }

    /// Store a serialized private key in the key store and send an update event
    /// to the application.
    pub fn set<F>(
        &self, id: impl Into<String> + Send + 'static, purpose: impl Into<String> + Send + 'static,
        data: impl Serialize + Send + 'static, make_event: F,
    ) where
        F: FnOnce(Result<(), KeyStoreError>) -> Ev + Send + Sync + 'static,
    {
        self.context.spawn({
            let context = self.context.clone();
            async move {
                let response = set(&context, id, purpose, data).await;
                context.update_app(make_event(response))
            }
        });
    }

    /// Store a serialized private key in the key store.
    pub async fn set_async(
        &self, id: impl Into<String>, purpose: impl Into<String>, data: impl Serialize,
    ) -> Result<(), KeyStoreError> {
        set(&self.context, id, purpose, data).await
    }

    /// Remove a serialized private key from the key store and send an update
    /// event to the application.
    pub fn delete<F>(
        &self, id: impl Into<String> + Send + 'static, purpose: impl Into<String> + Send + 'static,
        make_event: F,
    ) where
        F: FnOnce(Result<(), KeyStoreError>) -> Ev + Send + Sync + 'static,
    {
        self.context.spawn({
            let context: CapabilityContext<KeyStoreOperation, Ev> = self.context.clone();
            async move {
                let response = delete(&context, id, purpose).await;
                context.update_app(make_event(response))
            }
        });
    }

    /// Remove a serialized private key from the key store.
    pub async fn delete_async(
        &self, id: impl Into<String>, purpose: impl Into<String>,
    ) -> Result<(), KeyStoreError> {
        delete(&self.context, id, purpose).await
    }

    /// Generate a random secret suitable for key derivation and send an update
    /// event to the application.
    pub fn generate_secret<F>(&self, length: usize, make_event: F)
    where
        F: FnOnce(Result<Vec<u8>, KeyStoreError>) -> Ev + Send + Sync + 'static,
    {
        self.context.spawn({
            let context: CapabilityContext<KeyStoreOperation, Ev> = self.context.clone();
            async move {
                let response = generate_secret(&context, length).await;
                context.update_app(make_event(response))
            }
        });
    }

    /// Generate a random secret suitable for key derivation.
    pub async fn generate_secret_async(&self, length: usize) -> Result<Vec<u8>, KeyStoreError> {
        generate_secret(&self.context, length).await
    }
}

async fn get<Ev: 'static>(
    context: &CapabilityContext<KeyStoreOperation, Ev>, id: impl Into<String>,
    purpose: impl Into<String>,
) -> Result<KeyStoreEntry, KeyStoreError> {
    context
        .request_from_shell(KeyStoreOperation::Get {
            id: id.into(),
            purpose: purpose.into(),
        })
        .await
        .unwrap_get()
}

async fn set<Ev: 'static>(
    context: &CapabilityContext<KeyStoreOperation, Ev>, id: impl Into<String>,
    purpose: impl Into<String>, data: impl Serialize,
) -> Result<(), KeyStoreError> {
    let bytes = serde_json::to_vec(&data).map_err(|e| KeyStoreError::InvalidRequest {
        message: format!("failed to serialize key: {}", e),
    })?;
    context
        .request_from_shell(KeyStoreOperation::Set {
            id: id.into(),
            purpose: purpose.into(),
            data: bytes,
        })
        .await
        .unwrap_set()
}

async fn delete<Ev: 'static>(
    context: &CapabilityContext<KeyStoreOperation, Ev>, id: impl Into<String>,
    purpose: impl Into<String>,
) -> Result<(), KeyStoreError> {
    context
        .request_from_shell(KeyStoreOperation::Delete {
            id: id.into(),
            purpose: purpose.into(),
        })
        .await
        .unwrap_delete()
}

async fn generate_secret<Ev: 'static>(
    context: &CapabilityContext<KeyStoreOperation, Ev>, length: usize,
) -> Result<Vec<u8>, KeyStoreError> {
    context
        .request_from_shell(KeyStoreOperation::GenerateSecret { length })
        .await
        .unwrap_generate_secret()
}