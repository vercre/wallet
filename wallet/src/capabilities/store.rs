//! # Verifiable Credential Store Capability
use std::fmt::{self, Debug, Display};

use crux_core::capability::{CapabilityContext, Operation};
use crux_core::Capability;
use serde::{Deserialize, Serialize};
use thiserror::Error;

/// Some pre-defined store catalogs.
pub enum Catalog {
    /// Cedentials collection.
    Credential,
}

impl Display for Catalog {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Catalog::Credential => write!(f, "credential"),
        }
    }
}

/// Errors that can be returned by the Store capability.
#[derive(Clone, Debug, Deserialize, Serialize, Error, PartialEq, Eq)]
pub enum StoreError {
    /// Invalid request.
    #[error("invalid store request {message}")]
    InvalidRequest { message: String },

    /// The response from the shell capability was invalid.
    #[error("invalid store response {message}")]
    InvalidResponse { message: String },
}

/// An entry in the data store; a serialized credential or flow state.
///
/// `StoreEntry::None` is used to represent a missing entry in the store rather
/// than using an `Option` which is not supported across the FFI boundary in
/// Crux.
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub enum StoreEntry {
    /// No entry for a given ID.
    None,

    /// A serialized credential or flow state.
    Data(#[serde(with = "serde_bytes")] Vec<u8>),
}

impl From<Vec<u8>> for StoreEntry {
    fn from(bytes: Vec<u8>) -> Self {
        StoreEntry::Data(bytes)
    }
}

impl From<StoreEntry> for Option<Vec<u8>> {
    fn from(entry: StoreEntry) -> Option<Vec<u8>> {
        match entry {
            StoreEntry::None => None,
            StoreEntry::Data(bytes) => Some(bytes),
        }
    }
}

impl From<Option<Vec<u8>>> for StoreEntry {
    fn from(val: Option<Vec<u8>>) -> Self {
        match val {
            None => StoreEntry::None,
            Some(bytes) => StoreEntry::Data(bytes),
        }
    }
}

/// Supported operations for the Store capability.
#[derive(Clone, Serialize, Deserialize, PartialEq, Eq)]
pub enum StoreOperation {
    /// Save a serialized object to the store. Insert a new entry if the ID does
    /// not exist, otherwise update the existing entry.
    Save {
        catalog: String,
        id: String,
        #[serde(with = "serde_bytes")]
        data: Vec<u8>,
    },

    /// Get all serialized objects from the store.
    List { catalog: String },

    /// Remove the credential with the given ID from the store.
    Delete { catalog: String, id: String },
}

impl Debug for StoreOperation {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            StoreOperation::Save { catalog, id, data } => {
                let body = format!("<binary data - {} bytes>", data.len());
                f.debug_struct("Save")
                    .field("catalog", catalog)
                    .field("id", id)
                    .field("data", &body)
                    .finish()
            }
            StoreOperation::List { catalog } => {
                f.debug_struct("List").field("catalog", catalog).finish()
            }
            StoreOperation::Delete { catalog, id } => {
                f.debug_struct("Delete").field("catalog", catalog).field("id", id).finish()
            }
        }
    }
}

/// The result of an operation on the store.
///
/// Note: we cannot use Rust's `Result` and `Option` here because generics are
/// not supported across the FFI boundary in Crux.
#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq)]
pub enum StoreResult {
    /// The operation was successful.
    Ok { response: StoreResponse },

    /// The operation failed.
    Err { error: StoreError },
}

impl StoreResult {
    fn unwrap_save(self) -> Result<(), StoreError> {
        match self {
            StoreResult::Ok { response } => match response {
                StoreResponse::Saved => Ok(()),
                _ => {
                    panic!("unexpected response for Save operation: {:?}", response);
                }
            },
            StoreResult::Err { error } => Err(error.clone()),
        }
    }

    fn unwrap_list(self) -> Result<Vec<StoreEntry>, StoreError> {
        match self {
            StoreResult::Ok { response } => match response {
                StoreResponse::List { entries } => Ok(entries),
                _ => {
                    panic!("unexpected response for List operation: {:?}", response);
                }
            },
            StoreResult::Err { error } => Err(error.clone()),
        }
    }

    fn unwrap_delete(self) -> Result<(), StoreError> {
        match self {
            StoreResult::Ok { response } => match response {
                StoreResponse::Deleted => Ok(()),
                _ => {
                    panic!("unexpected response for Delete operation: {:?}", response);
                }
            },
            StoreResult::Err { error } => Err(error.clone()),
        }
    }
}

/// The possible responses from the Store capability.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum StoreResponse {
    /// The result of a save operation.
    Saved,

    /// The result of a list operation.
    List { entries: Vec<StoreEntry> },

    /// The result of a delete operation.
    Deleted,
}

impl Operation for StoreOperation {
    type Output = StoreResult;
}

/// The type used to implement the capability.
pub struct Store<Ev> {
    context: CapabilityContext<StoreOperation, Ev>,
}

impl<Ev> Capability<Ev> for Store<Ev> {
    type MappedSelf<MappedEv> = Store<MappedEv>;
    type Operation = StoreOperation;

    fn map_event<F, NewEv>(&self, f: F) -> Self::MappedSelf<NewEv>
    where
        F: Fn(NewEv) -> Ev + Send + Sync + 'static,
        Ev: 'static,
        NewEv: 'static + Send,
    {
        Store::new(self.context.map_event(f))
    }

    #[cfg(feature = "typegen")]
    fn register_types(generator: &mut crux_core::typegen::TypeGen) -> crux_core::typegen::Result {
        generator.register_type::<StoreResponse>()?;
        generator.register_type::<StoreError>()?;
        generator.register_type::<StoreEntry>()?;
        generator.register_type::<Self::Operation>()?;
        generator.register_type::<<Self::Operation as Operation>::Output>()?;
        Ok(())
    }
}

impl<Ev> Clone for Store<Ev> {
    fn clone(&self) -> Self {
        Self {
            context: self.context.clone(),
        }
    }
}

impl<Ev> Store<Ev>
where
    Ev: 'static,
{
    /// Create a new instance of the Store capability.
    pub fn new(context: CapabilityContext<StoreOperation, Ev>) -> Self {
        Self { context }
    }

    /// Save a serialized credential to the store.
    pub fn save<F>(
        &self, catalog: impl Into<String> + Send + 'static, id: impl Into<String> + Send + 'static,
        data: impl Serialize + Send + 'static, make_event: F,
    ) where
        F: FnOnce(Result<(), StoreError>) -> Ev + Send + Sync + 'static,
    {
        let context = self.context.clone();
        self.context.spawn({
            async move {
                let response = save(&context, catalog, id, data).await;
                context.update_app(make_event(response))
            }
        });
    }

    /// Save a serialized credential to the store while in an async context.
    ///
    /// This can be used in a higher-order capability such as a `Provider` for
    /// `vercre-holder`.
    pub async fn save_async(
        &self, catalog: impl Into<String>, id: impl Into<String>, data: impl Serialize,
    ) -> Result<(), StoreError> {
        save(&self.context, catalog, id, data).await
    }

    /// Get all serialized credentials from the store.
    pub fn list<F>(&self, catalog: impl Into<String> + Send + 'static, make_event: F)
    where
        F: FnOnce(Result<Vec<StoreEntry>, StoreError>) -> Ev + Send + Sync + 'static,
    {
        self.context.spawn({
            let context = self.context.clone();
            async move {
                let response = list(&context, catalog).await;
                context.update_app(make_event(response))
            }
        });
    }

    /// Get all serialized credentials from the store while in an async context.
    ///
    /// This can be used in a higher-order capability such as a `Provider` for
    /// `vercre-holder`.
    pub async fn list_async(
        &self, catalog: impl Into<String>,
    ) -> Result<Vec<StoreEntry>, StoreError> {
        list(&self.context, catalog).await
    }

    /// Remove the credential with the given ID from the store.
    pub fn delete<F>(
        &self, catalog: impl Into<String> + Send + 'static, id: impl Into<String> + Send + 'static,
        make_event: F,
    ) where
        F: FnOnce(Result<(), StoreError>) -> Ev + Send + Sync + 'static,
    {
        self.context.spawn({
            let context = self.context.clone();
            async move {
                let response = delete(&context, catalog, id).await;
                context.update_app(make_event(response))
            }
        });
    }

    /// Remove the credential with the given ID from the store while in an async
    /// context.
    ///
    /// This can be used in a higher-order capability such as a `Provider` for
    /// `vercre-holder`.
    pub async fn delete_async(
        &self, catalog: impl Into<String>, id: impl Into<String>,
    ) -> Result<(), StoreError> {
        delete(&self.context, catalog, id).await
    }
}

async fn save<Ev: 'static>(
    context: &CapabilityContext<StoreOperation, Ev>, catalog: impl Into<String>,
    id: impl Into<String>, data: impl Serialize,
) -> Result<(), StoreError> {
    let bytes = serde_json::to_vec(&data).map_err(|e| StoreError::InvalidRequest {
        message: format!("failed to serialize data: {}", e),
    })?;
    context
        .request_from_shell(StoreOperation::Save {
            catalog: catalog.into(),
            id: id.into(),
            data: bytes,
        })
        .await
        .unwrap_save()
}

async fn list<Ev: 'static>(
    context: &CapabilityContext<StoreOperation, Ev>, catalog: impl Into<String>,
) -> Result<Vec<StoreEntry>, StoreError> {
    context
        .request_from_shell(StoreOperation::List {
            catalog: catalog.into(),
        })
        .await
        .unwrap_list()
}

async fn delete<Ev: 'static>(
    context: &CapabilityContext<StoreOperation, Ev>, catalog: impl Into<String>,
    id: impl Into<String>,
) -> Result<(), StoreError> {
    context
        .request_from_shell(StoreOperation::Delete {
            catalog: catalog.into(),
            id: id.into(),
        })
        .await
        .unwrap_delete()
}
