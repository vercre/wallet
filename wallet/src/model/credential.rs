//! Credential sub-app state.

use serde::{Deserialize, Serialize};
use vercre_holder::credential::Credential;

/// Application state for the credential sub-app.
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[allow(clippy::module_name_repetitions)]
pub struct CredentialState {
    /// Currently selected credential.
    pub id: Option<String>,

    /// Credentials stored in the wallet.
    pub credentials: Vec<Credential>,
}

impl CredentialState {
    /// Create a new credential state.
    #[must_use]
    pub fn init() -> Self {
        let json = include_bytes!("credentials.json");
        let credentials: Vec<Credential> = serde_json::from_slice(json).expect("should deserialize");
        Self {
            id: None,
            credentials,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_init() {
        let state = CredentialState::init();
        assert_eq!(state.credentials.len(), 2);
    }
}
