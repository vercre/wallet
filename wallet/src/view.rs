//! # View Model
//! 
pub mod credential;

use credential::CredentialView;
use serde::{Deserialize, Serialize};

use super::Aspect;

/// View model for the wallet application.
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
pub struct ViewModel {
    /// Which aspect of the application is currently active.
    pub active_view: Aspect,

    /// Credential view model.
    pub credential_view: CredentialView,

    // --- TODO: Remove ---
    pub text: String,
    pub confirmed: bool,
    // --------------------
}
