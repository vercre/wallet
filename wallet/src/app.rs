//! This module contains the core application fabric for the wallet, including
//! the model, events, and effects that drive the application.

use crux_core::render::Render;
use crux_http::Http;
use crux_kv::KeyValue;
use serde::{Deserialize, Serialize};

use crate::capabilities::key::KeyStore;
use crate::capabilities::sse::ServerSentEvents;
use crate::capabilities::store::{Catalog, Store, StoreEntry, StoreError};
use crate::model::Model;
use crate::provider::Provider;
use crate::view::ViewModel;

/// Aspect of the application.
///
/// This allows the UI navigation to be reactive: controlled in response to the
/// user's actions.
#[derive(Clone, Default, Debug, Deserialize, Serialize, PartialEq, Eq)]
pub enum Aspect {
    /// Display and deletion of credentials stored in the wallet.
    #[default]
    CredentialList,

    /// Display of a single credential.
    CredentialDetail,

    /// Trigger a credential issuance using an offer QR code.
    IssuanceScan,

    /// View the offer details to decide whether or not to proceed with
    /// issuance.
    IssuanceOffer,

    /// Trigger a credential verification using a presentation request QR code.
    PresentationScan,

    /// The application is in an error state.
    Error,
}

/// Events that can be sent to the wallet application.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub enum Event {
    /// Error event is emitted by the core when an error occurs.
    #[serde(skip)]
    Error(String),

    //--- Credential events ----------------------------------------------------
    /// Event emitted by the shell when the app first loads.
    Ready,

    /// Event emitted by the shell to select a credential from the list of
    /// stored credentials for detailed display.
    SelectCredential(String),

    /// Event emitted by the shell to delete a credential from the wallet.
    DeleteCredential(String),

    /// Event emitted by the core when the store capability has loaded
    /// credentials.
    #[serde(skip)]
    CredentialsLoaded(Result<Vec<StoreEntry>, StoreError>),

    /// Event emitted by the core when the store capability has stored a
    /// credential.
    #[serde(skip)]
    CredentialStored(Result<(), StoreError>),

    /// Event emitted by the core when the store capability has deleted a
    /// credential.
    #[serde(skip)]
    CredentialDeleted(Result<(), StoreError>),

    //--- Issuance events ------------------------------------------------------
    /// Event emitted by the shell when the user wants to scan an issuance offer
    /// QR code.
    ScanIssuanceOffer,

    /// Event emitted by the shell when the user scans an offer QR code.
    IssuanceOffer(String),
}

#[cfg_attr(feature = "typegen", derive(crux_core::macros::Export))]
#[derive(crux_core::macros::Effect)]
pub struct Capabilities {
    pub render: Render<Event>,
    pub http: Http<Event>,
    pub key_store: KeyStore<Event>,
    pub kv: KeyValue<Event>,
    pub sse: ServerSentEvents<Event>,
    pub store: Store<Event>,
}

#[derive(Default)]
pub struct App;

impl crux_core::App for App {
    type Capabilities = Capabilities;
    type Event = Event;
    type Model = Model;
    type ViewModel = ViewModel;

    fn update(&self, msg: Self::Event, model: &mut Self::Model, caps: &Self::Capabilities) {
        match msg {
            Event::Error(e) => {
                model.error(e);
                caps.render.render();
            }
            Event::Ready => {
                model.ready();
                caps.store.list("credential", Event::CredentialsLoaded);
                caps.render.render();
            }
            Event::SelectCredential(id) => {
                model.select_credential(id);
                caps.render.render();
            }
            Event::DeleteCredential(id) => {
                caps.store.delete("credential", id, Event::CredentialDeleted);
            }
            Event::CredentialsLoaded(Ok(entries)) => {
                model.credentials_loaded(entries);

                // TODO: Remove this try-out code
                // if entries.len() < 1 {
                //     let json = include_bytes!("model/credentials.json");
                //     let credentials: Vec<Credential> =
                //         serde_json::from_slice(json).expect("should deserialize");
                //     let credential = credentials[entries.len()].clone();
                //     let data = serde_json::to_vec(&credential).expect("should serialize");
                //     caps.store.save("credential".into(), credential.id.clone(), data,
                // Event::CredentialStored); }
                // -------------------------------

                caps.render.render();
            }
            Event::CredentialsLoaded(Err(error)) => {
                model.error(error.to_string());
                caps.render.render();
            }
            Event::CredentialStored(Ok(())) => {
                caps.store.list(Catalog::Credential.to_string(), Event::CredentialsLoaded);
            }
            Event::CredentialStored(Err(error)) => {
                model.error(error.to_string());
                caps.render.render();
            }
            Event::CredentialDeleted(Ok(())) => {
                model.delete_credential();
                caps.store.list(Catalog::Credential.to_string(), Event::CredentialsLoaded);
                caps.render.render();
            }
            Event::CredentialDeleted(Err(error)) => {
                model.error = Some(error.to_string());
                caps.render.render();
            }
            Event::ScanIssuanceOffer => {
                model.scan_issuance_offer();
                caps.render.render();
            }
            Event::IssuanceOffer(encoded_offer) => {
                let provider = Provider::new(
                    caps.http.clone(),
                    caps.key_store.clone(),
                    caps.kv.clone(),
                    caps.store.clone(),
                );
                model.issuance_offer(&provider, &encoded_offer);
                caps.render.render();
            }
        }
    }

    fn view(&self, model: &Self::Model) -> Self::ViewModel {
        Self::ViewModel {
            active_view: model.active_view.clone(),
            credential_view: model.credential.clone().into(),
        }
    }
}
