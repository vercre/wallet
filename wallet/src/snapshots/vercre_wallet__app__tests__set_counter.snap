---
source: wallet/src/app.rs
assertion_line: 250
expression: app.view(&model)
snapshot_kind: text
---
active_view: CredentialList
credential_view:
  id: ~
  credentials: []
text: "1 (2023-01-01 00:00:00 UTC)"
confirmed: true
